content: draft: address issues in threats.md (#1310)
slsa-framework/slsa: docs/spec/draft/threats.md

</details>

#### (B4) Render logs ineffective

> <details><summary>Forge change metadata</summary>

Reviewer: Can we change the heading to be something more representative of the summary? I assume that this is logs for incident response, etc. but to me, simple statements of logs are often build logs.

Author: Do you have a suggestion?

Reviewer: ```
#### (B4) Render commit metadata
```
?

Author: How about "Render change metadata ineffective"

"commit" -> "change" to stay consistent (and somewhat platform agnostic)

Author: Done.  Let me know if you don't like the new thing. :)

</details>

#### (B4) Render logs ineffective

<details><summary>Forge change metadata</summary>

*Threat:* Forge the change metadata to alter attribution, timestamp, or
discoverability of a change.

> *Mitigation:* Source control platform strongly authenticates actor identity,

Reviewer 2: I recommend dropping this item, saying that slsa doesn't support until a future track, or mentioning that consumers have to have their own way to meaningfully identify contributors?

> *Mitigation:* Source control platform strongly authenticates actor identity,

Author: Done.  Now just saying SLSA doesn't address this yet.

*Example:* MyPackage is built on Awesome Builder under the project "mypackage".
Adversary is an administrator of the "mypackage" project. Awesome Builder allows
administrators to debug build machines via SSH. An adversary uses this feature
to alter a build in progress. Solution: Build L3 requires provenance to be
resistant to forgery by tenants, so the attacker would be forbidden from

> altering a build in progress.

Reviewer: Altering a build is different from forging provenance. This is also not an action by a tenant but an admin. I initially thought that this related to isolation, but that is more scoped for preventing parallel builds from affecting each other.

I feel like this falls more into the "future" operations track. How can a build platform be run securely so that admin users have the least possible impact.

Author: So, now that you call this out I think the example differs from the 'threat' and 'mitigation'.

The threat and mitigation are talking about an attacker gaining ownership of a _project_.  The example then changes terminology to talk about 'administrator' access which makes us think it's talking about administration of the entire build platform, but it is really just supposed to be talking about _the project itself_.

Re "forging", the need for the provenance to be complete comes from the high level "Provenance is Unforgeable" requirement.  Let me replace that with something more specific.

So I think we really want something like

> *Example:* MyPackage is built on Awesome Builder under the project "mypackage".
Adversary is an _owner_ of the "mypackage" project. Awesome Builder allows
_owners_ to debug the _build environment_ via SSH. An adversary uses this feature
to alter a build in progress. Solution: Build L3 requires _complete_ provenance, the
attackers access and/or actions within the SSH would be enumerated within the
external parameters.  The updated external parameters will not match the declared
expectations causing verification to fail.

I'd agree that any attacks by the Build Platform administrator would definitely fall under some new tracks (like Build Environment).

SG?

Author: Fixed.  Thanks for the catch.

the source file. Adversary runs a malicious build that creates a "poisoned"
cache entry with a falsified key, meaning that the value wasn't really produced
from that source. A subsequent build then picks up that poisoned cache entry.

> Solution: Builder uses a separate cache for each build.

Reviewer 2: ```suggestion
Solution: Builder.id uses a separate cache for each build.
```
Right? there's no way for actions as a product to use a separate cache for all builds, but you might want something like that for specific configurations.

Reviewer 2: potentially in this case, the solution is: "for critical builds, do not use a cache."

Author: Builder.id isn't a component in the [Build model](https://slsa.dev/spec/draft/terminology#build-model), but it does [_refer to_](https://slsa.dev/spec/v1.0/provenance#builder) the 'Build Platform'.  In this doc we're always using 'Builder' as a shorthand for 'Build Platform'.

In the case of GitHub Actions, I don't think we'd consider GitHub Actions the "build platform" instead it's a component of that platform.  Much like for an [SCS](https://slsa.dev/spec/draft/source-requirements#definitions:~:text=Source%20Control%20System%20(SCS)) the system is a collection of tools and services that people build on top of to meet the requirements.  For GHA and SLSA builds it's really the specific workflow that brings everything together and I think _it_ can decide how caches are or aren't used?  So I think we're fine?

> potentially in this case, the solution is: "for critical builds, do not use a cache."

I'd rather not start talking about 'critical builds'.  Everywhere else in this page we just talk about 'builds'

*Example:* An adversary submits malicious code to the MyPackage GitHub repo,
builds from that revision, then does a force push to erase that revision from

> history (or requests that GitHub delete the repo.) This would make the revision

Reviewer 2: ```suggestion
history deletes the repo. This would make the revision
```

unavailable for inspection. Solution: Verifier rejects packages built from

> source repos that do not have strong retention guarantees.

Reviewer 2: who makes the guarantee in this case?

we could also say "verifier rejects the package because it lacks positive attestations that consumed source is still available."

Author: I'm trying to get away from talking about attestations in the source threats section because those attestations don't exist yet (they're part of the source track).

> "who makes the guarantee in this case?"

This would be the source control platform.  I don't know how many actually do this.

I'd also be happy saying SLSA doesn't address this threat yet, which would be simpler.

Note that I don't think _anything_ we're planning in the source track would prevent a repo owner from deleting that repo.  Whenever we address these things we'd probably want to provide separate examples/mitigations.

Reviewer 3: > I'd also be happy saying SLSA doesn't address this threat yet, which would be simpler.

I support this approach.

Author: Done.

*Mitigation:* Require cryptographically secure hash functions for commit

> checksums and provenance subjects, such as SHA-256.

Reviewer 2: ```suggestion
IDs and provenance subjects, such as SHA-256.
```

> *Examples:* Attacker crafts a malicious executable with the same MD5 hash as a

Author: I was just trying to make the example as meaningful as possible.  I can replace executable with file though. That's fine.

Generally I think commit IDs might be misplaced here.  Why are they special?  Even once we have the source track we'll still really just care about what's in the provenance.

Maybe we just drop 'commit IDs' and only refer to provenance subjects?

Another potential issue (that I'm not sure I want to get into) is that sometimes there are [use cases](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md#use-cases-for-non-cryptographic-immutable-digests) for non-cryptographic identifiers in digests and this would seem to preclude that (but not really since this isn't the requirements).  If we wanted to clarify _that_ we'd have to say something like "*Mitigation:* Choose **secure** algorithms when using cryptographic digests"

> *Examples:* Attacker crafts a malicious executable with the same MD5 hash as a

Reviewer 3: > Maybe we just drop 'commit IDs' and only refer to provenance subjects?

I'd drop commit IDs from here. I know this is "only" the threats page and not a requirement for a track, but I wonder if some folks will get confused and think that SLSA requires them to start using git with SHA-256 identifiers.

> Generally I think commit IDs might be misplaced here. Why are they special? Even once we have the source track we'll still really just care about what's in the provenance.

I think the source provenance or VSA subject might still refer to a revision using `gitCommit` or so as the digest type, which ends up being SHA-1? (At least that's what we're doing for the VSA PoC in gittuf)

> *Examples:* Attacker crafts a malicious executable with the same MD5 hash as a

Author: Yes it does, that's a good point!
